compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/CollectionStubMethodLowering.kt

```diff
@@ -0,0 +1,274 @@
+        return buildFun {
+            name = function.name
+            returnType = function.returnType.substitute(typeMap)
+            visibility = function.visibility
+            origin = IrDeclarationOrigin.IR_BUILTINS_STUB
+            modality = Modality.OPEN
+        }.apply {
+            // Replace Function metadata with the data from class
+            // Add the abstract function symbol to stub function for bridge lowering
+            overriddenSymbols.add(function.symbol)
+            parent = irClass
+            dispatchReceiverParameter = function.dispatchReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            extensionReceiverParameter = function.extensionReceiverParameter?.copyWithTypeErasure(this, typeMap)
+            function.valueParameters.mapIndexed { i, param ->
+                valueParameters.add(i, param.copyWithTypeErasure(this, typeMap))
+            }
+            val exception = context.getTopLevelClass(FqName("java.lang.UnsupportedOperationException"))
+            // Function body consist only of throwing UnsupportedOperationException statement
+            body = context.createIrBuilder(function.symbol).irBlockBody {
+                +irThrow(
+                    irCall(
+                        exception.owner.constructors.single {
+                            it.valueParameters.size == 1 && it.valueParameters.single().type.isNullableString()
+                        }
+                    ).apply {
+                        putValueArgument(0, irString("Operation is not supported for read-only collection"))
+                    }
+                )
+            }
+        }
+    }
+
+    // Type erasure for value parameters, with type substitution
+    private fun IrValueParameter.copyWithTypeErasure(target: IrSimpleFunction, typeMap: Map<IrTypeParameterSymbol, IrType>)
+            : IrValueParameter {
+        val descriptor = WrappedValueParameterDescriptor(this.descriptor.annotations)
+        return IrValueParameterImpl(
+            UNDEFINED_OFFSET, UNDEFINED_OFFSET,
+            IrDeclarationOrigin.IR_BUILTINS_STUB,
+            IrValueParameterSymbolImpl(descriptor),
+            name,
+            index,
+            type.substitute(typeMap).eraseTypeParameters(),
+            varargElementType?.substitute(typeMap)?.eraseTypeParameters(),
+            isCrossinline,
+            isNoinline
+        ).apply {
+            descriptor.bind(this)
+            parent = target
+        }
+    }
+
+    // Create a map for type parameters between source class (Mutable Collection classes) to
+    // target class (class currently in lowering phase), this map is later used for substituting type parameters in generated functions
+    private fun createTypeMap(collectionClass: IrClass, mutableClass: IrClass, targetClass: IrClass): Map<IrTypeParameterSymbol, IrType> {
```

Comment:
I'd prefer to avoid relying on functions with specific hardcoded names to find out which argument is used for the collection's type parameter. I think the code would be easier and more future-proof if we constructed the list of all (substituted) supertypes of `targetClass`, picked the one supertype whose class is `mutableClass`, and extracted all type arguments from it. Pretty much what the existing code in the old backend does here: https://github.com/JetBrains/kotlin/blob/6dbbf0bb7250b1c8ca11a1a925eb3fe468f2e89a/compiler/backend/src/org/jetbrains/kotlin/codegen/CollectionStubMethodGenerator.kt#L97

Reply:
That's also my concern for this approach as well. Regarding the code you pointed, when I looked into that code during initial investigation, I feel the old logic is heavily relying on descriptor, which is what we want to get away with new back end, therefore I am not sure if we want to follow that logic.

Reply:
Right. My suggestion was not to use that code directly, but reimplement it for `IrType`

Reply:
@udalov Can you take a look and let me know if latest revision is what you are expecting? Thanks!